graph_worker.cpp

```diff
@@ -33,31 +32,24 @@ GraphWorker::GraphWorker(int _id, Graph *_graph, BufferTree *_bf) :
 }
 
 GraphWorker::~GraphWorker() {
+	if(id == 0)
+		bf->bypass_wait(); // to avoid race condition on shutdown. Tell threads to recheck wait
+	
 	// when a graph worker is deleted we must wait for the associated
 	// thread to finish its work. See stopWorkers and doWork
 	thr.join();
 }
 
 void GraphWorker::doWork() {
+	data_ret_t data;
 	while(true) {
-		std::unique_lock<std::mutex> queue_unique(bf->queue_lock);
-		bf->queue_cond.wait(queue_unique, [this]{return (!bf->work_queue.empty() || shutdown);});
-
-		if (!bf->work_queue.empty()) {
-			work_t task = bf->work_queue.front();
-			bf->work_queue.pop();
-			queue_unique.unlock();  // unlock
-			// printf("Worker %d handling updates for node %llu\n", id, node);
-			std::pair<Node, std::vector<Node>> data = bf->get_data(task);
-			graph->batch_update(data.first, data.second);
-		}
-		else queue_unique.unlock(); // unlock
+		// call get_data which will handle waiting on the queue
+		// and will enforce locking.
+		bool valid = bf->get_data(data);
 
-		// doesn't really matter if this part isn't thread safe I believe
-		// only reading and never should read empty when there is more to insert
-		if (shutdown && bf->work_queue.empty()) {// if recieved shutdown and there's no more work to do
-			printf("Thread %i done and exiting\n", id);
+		if (valid)
+			graph->batch_update(data.first, data.second);
```

Comment:
wait for implicit/explicit sync impl

Reply:
What?

Reply:
since `batch_update` isn't synchronized, multiple workers could try to access a sketch concurrently. we need to sync access, either implicitly (within `sketch::operator+=`) or explicitly.

Reply:
Oh yes. Absolutely, however this change (the circular queue) isn't what will make that necessary. The basement nodes are what will do that.